src/librbd/deep_copy/ObjectCopyRequest.cc

```diff
@@ -0,0 +1,684 @@
+          } else if (e.offset < prev_end_size) {
+            auto len = prev_end_size - e.offset;
+            ldout(m_cct, 20) << "src object not exist and extent " << e.offset
+                             << "~" << e.length << " intersects object prev end"
+                             << " " << prev_end_size << ", inserting zero "
+                             << dst_object_offset << "~" << len << dendl;
+            m_zero_interval[end_src_snap_id].insert(dst_object_offset, len);
+            m_end[end_src_snap_id] = std::max(dst_object_offset + len,
+                                              m_end[end_src_snap_id]);
+          }
+        }
+        dst_object_offset += e.length;
+      }
+    }
+
+    prev_end_size = end_size;
+    prev_exists = exists;
+    start_src_snap_id = end_src_snap_id;
+  }
+
+  for (auto &it : m_read_ops) {
+    m_read_snaps.push_back(it.first);
+  }
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::merge_write_ops() {
+  ldout(m_cct, 20) << dendl;
+
+  for (auto &it : m_zero_interval) {
+    m_dst_zero_interval[it.first].insert(it.second);
+  }
+
+  for (auto &it : m_end) {
+    m_dst_object_end[it.first] = std::max(m_dst_object_end[it.first],
+                                          it.second);
+  }
+
+  for (auto &it : m_read_ops) {
+    auto src_snap_seq = it.first.first;
+    auto &copy_ops = it.second;
+    for (auto &copy_op : copy_ops) {
+      uint64_t src_offset = copy_op.src_offset;
+      uint64_t dst_offset = copy_op.dst_offset;
+      for (auto &e : copy_op.src_extent_map) {
+        uint64_t zero_len = e.first - src_offset;
+        if (zero_len > 0) {
+          ldout(m_cct, 20) << "src_snap_seq=" << src_snap_seq
+                           << ", inserting zero " << dst_offset << "~"
+                           << zero_len << dendl;
+          m_dst_zero_interval[src_snap_seq].insert(dst_offset, zero_len);
+          src_offset += zero_len;
+          dst_offset += zero_len;
+        }
+        copy_op.dst_extent_map[dst_offset] = e.second;
+        src_offset += e.second;
+        dst_offset += e.second;
+      }
+      if (dst_offset < copy_op.dst_offset + copy_op.length) {
+        uint64_t zero_len = copy_op.dst_offset + copy_op.length - dst_offset;
+        ldout(m_cct, 20) << "src_snap_seq=" << src_snap_seq
+                         << ", inserting zero " << dst_offset << "~"
+                         << zero_len << dendl;
+        m_dst_zero_interval[src_snap_seq].insert(dst_offset, zero_len);
+      } else {
+        assert(dst_offset == copy_op.dst_offset + copy_op.length);
+      }
+      m_write_ops[src_snap_seq].emplace_back(std::move(copy_op));
+    }
+  }
+}
+
+template <typename I>
+void ObjectCopyRequest<I>::compute_zero_ops() {
+  ldout(m_cct, 20) << dendl;
+
+  bool fast_diff;
+  {
+    RWLock::RLocker snap_locker(m_dst_image_ctx->snap_lock);
+    fast_diff = m_dst_image_ctx->test_features(RBD_FEATURE_FAST_DIFF,
```

Comment:
Nit: you can use the version that doesn't have a lock param to avoid the clause and allow you to directly define and initialize ```fast_diff``` in one line of code

Reply:
I was avoiding to mix two versions of `test_features` in one file as it would complicate unit tests, where I would need to support mocks for both versions. In this file `test_features` is also used in `send_update_object_map`. It looks I can change it too to use he version that doesn't have a lock param, so I am going to update this as you are suggesting.

Reply:
Make sense -- no worries if it's a pain.